### DISCUSSION TITLE
Fetch data from route handlers in server components

### DISCUSSION THREAD

**Discussion by tomdohnal** [vercel/next.js#48448] [Help]
### Summary

Hello, 

I'm trying to fetch data in server components from route handlers.

However, when I build the app, I get a `TypeError: fetch failed` error (`code: 'ECONNREFUSED'`).

Ideally, I'd put the server code in the server components directly (in my case, that would be some Prisma DB calls).

However, by doing that, I'd like the fine-grained control over caching I get when I use `fetch(...)`.

Do you have any recommendations?

**Comment by devjiwonchoi** [comment]
The "API route" of your localhost cannot be established during the build time.

### Work Around

> Which could be a bit annoying...

#### Option 1

1.  Mock your `route.ts` with a simple express app
2. Run the express server during build time

#### Option 2

1. Deploy your Next.js app with your `route.ts`, without any dependant pages.
2. Redeploy with your pages

**Note that `route.ts` are currently not meant to be an `API Server`, but rather a `route handler`.**

**Comment by Nadeera3784** [comment]
Hi @tomdohnal ,
You can simply use useServerState.
